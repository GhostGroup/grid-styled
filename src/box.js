import React, { Component } from "react";
import styled from "styled-components";
import { space, width, responsiveStyle } from "styled-system";
import { number, string, array, oneOfType } from "prop-types";
import propTypes from "./prop-types";
const responsivePropType = oneOfType([number, string, array]);
const flex = responsiveStyle("flex");
const order = responsiveStyle("order");

const BoxTypeMap = {};

class Box extends Component {
  static propTypes = {
    ...propTypes,
    flex: responsivePropType,
    order: responsivePropType
  };

  static defaultProps = {
    is: "div"
  };

  render() {
    const { is, ...rest } = this.props;
    if (!BoxTypeMap[is]) {
      BoxTypeMap[is] = styled(is)(
        [],
        { boxSizing: "border-box" },
        width,
        space,
        flex,
        order
      );
    }
    const Comp = BoxTypeMap[is];
    return <Comp {...rest} />;
  }
}

export default Box;
