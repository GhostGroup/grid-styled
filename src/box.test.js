import React from "react";
import styled from "styled-components";
import { shallow, mount } from "enzyme";
import Box from "./box";

describe("box", () => {
  it("renders", () => {
    const tree = shallow(<Box m={2} px={3} />);
    expect(tree).toMatchSnapshot();
  });

  it("renders with props", () => {
    const tree = mount(<Box m={[1, 2]} px={[1, 2]} w={1} flex="1 1 auto" />);
    expect(tree).toMatchSnapshot();
  });

  it("can be extended", () => {
    const MyBox = styled(Box)`
      background-color: ${props => (props.isActive ? "blue" : "red")};
    `;
    const tree = mount(
      <MyBox m={[1, 2]} px={[1, 2]} w={1} flex="1 1 auto" isActive />
    );
    expect(tree).toMatchSnapshot();
  });

  it('can use a different component with the "is" prop', () => {
    const tree = mount(
      <Box m={2} px={3} is="section">
        Test
      </Box>
    );
    expect(tree).toMatchSnapshot();
  });

  it("renders the same component when props change", () => {
    class Component extends React.Component {
      render() {
        return (
          <Box {...this.props}>
            <div ref="childRef" />
          </Box>
        );
      }
    }
    const tree = mount(<Component />);
    const childRef = tree.ref("childRef");
    tree.setProps({ value: "a new value" });
    expect(tree.ref("childRef")).toBe(childRef);
  });
});
