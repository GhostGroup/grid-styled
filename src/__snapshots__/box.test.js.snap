// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`box can be extended 1`] = `
.c1 {
  box-sizing: border-box;
  width: 100%;
  margin: 8px;
  padding-left: 8px;
  padding-right: 8px;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
}

.c0 {
  background-color: blue;
}

@media screen and (min-width:40em) {
  .c1 {
    margin: 16px;
    padding-left: 16px;
    padding-right: 16px;
  }
}

<Styled(Box)
  flex="1 1 auto"
  isActive={true}
  m={
    Array [
      1,
      2,
    ]
  }
  px={
    Array [
      1,
      2,
    ]
  }
  w={1}
>
  <Box
    className="c0"
    flex="1 1 auto"
    is="div"
    isActive={true}
    m={
      Array [
        1,
        2,
      ]
    }
    px={
      Array [
        1,
        2,
      ]
    }
    w={1}
  >
    <styled.div
      className="c0"
      flex="1 1 auto"
      isActive={true}
      m={
        Array [
          1,
          2,
        ]
      }
      px={
        Array [
          1,
          2,
        ]
      }
      w={1}
    >
      <div
        className="c0 c1"
      />
    </styled.div>
  </Box>
</Styled(Box)>
`;

exports[`box can use a different component with the "is" prop 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 16px;
  padding-left: 32px;
  padding-right: 32px;
}

<Box
  is="section"
  m={2}
  px={3}
>
  <styled.section
    m={2}
    px={3}
  >
    <section
      className="c0"
    >
      Test
    </section>
  </styled.section>
</Box>
`;

exports[`box renders 1`] = `
<styled.div
  m={2}
  px={3}
/>
`;

exports[`box renders with props 1`] = `
.c0 {
  box-sizing: border-box;
  width: 100%;
  margin: 8px;
  padding-left: 8px;
  padding-right: 8px;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
}

@media screen and (min-width:40em) {
  .c0 {
    margin: 16px;
    padding-left: 16px;
    padding-right: 16px;
  }
}

<Box
  flex="1 1 auto"
  is="div"
  m={
    Array [
      1,
      2,
    ]
  }
  px={
    Array [
      1,
      2,
    ]
  }
  w={1}
>
  <styled.div
    flex="1 1 auto"
    m={
      Array [
        1,
        2,
      ]
    }
    px={
      Array [
        1,
        2,
      ]
    }
    w={1}
  >
    <div
      className="c0"
    />
  </styled.div>
</Box>
`;
